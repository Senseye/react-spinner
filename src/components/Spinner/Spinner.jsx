import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

class Spinner extends PureComponent {
  constructor(props) {
    super(props);
    this.setSVGParams();
  }

  setSVGParams() {
    const {
      circleAttributes: {
        radius,
        circleStrokeWidth,
      },
    } = this.props;

    this.circleLength = 2 * Math.PI * radius;
    this.svgSize = (radius * 2) + circleStrokeWidth;
    this.midPoint = radius + (circleStrokeWidth / 2);
    this.fontSize = this.svgSize / 3.75;
    this.textY = this.midPoint + (this.fontSize / 3);
  }

  isCompleted() {
    const { actualProgress, finalProgressValue } = this.props;
    return actualProgress >= finalProgressValue;
  }

  renderTextProgress() {
    const { actualProgress, finalProgressValue } = this.props;
    const { midPoint, fontSize, textY } = this;
    const percentageCompleted = Math.floor((actualProgress * 100) / finalProgressValue);
    return (
      <text x={midPoint} y={textY} textAnchor="middle" fontSize={fontSize}>
        {percentageCompleted}
      </text>
    );
  }

  renderCircleRotation() {
    const { circleAttributes: { rotationSpeed } } = this.props;
    const { midPoint } = this;

    return !this.isCompleted() && (
      <animateTransform
        attributeName="transform"
        attributeType="XML"
        type="rotate"
        from={`360 ${midPoint} ${midPoint}`}
        to={`0 ${midPoint} ${midPoint}`}
        dur={rotationSpeed}
        repeatCount="indefinite"
      />
    );
  }

  renderProgressCircle() {
    const {
      actualProgress,
      finalProgressValue,
      circleAttributes: {
        radius,
        circleStrokeWidth,
        circleProgressColor,
      },
    } = this.props;

    const {
      circleLength,
      midPoint,
    } = this;

    const offset = circleLength * (1 - (actualProgress / finalProgressValue));

    return actualProgress > 0 ? (
      <circle
        fill="none"
        cx={midPoint}
        cy={midPoint}
        r={radius}
        stroke={circleProgressColor}
        strokeWidth={circleStrokeWidth}
        strokeDasharray={circleLength}
        strokeDashoffset={offset}
      >
        { this.renderCircleRotation() }
      </circle>
    ) : null;
  }

  render() {
    const {
      circleAttributes: {
        radius,
        circleStrokeWidth,
        circleColor,
      },
    } = this.props;

    const { svgSize, midPoint } = this;

    return (
      <svg width={svgSize} height={svgSize}>
        <circle cx={midPoint} cy={midPoint} r={radius} fill="none" stroke={circleColor} strokeWidth={circleStrokeWidth} />
        { this.renderProgressCircle() }
        { this.renderTextProgress() }
      </svg>
    );
  }
}

Spinner.propTypes = {
  actualProgress: PropTypes.number.isRequired,
  finalProgressValue: PropTypes.number,
  circleAttributes: PropTypes.shape({
    radius: PropTypes.number,
    circleStrokeWidth: PropTypes.number,
    circleColor: PropTypes.string,
    circleProgressColor: PropTypes.string,
    rotationSpeed: PropTypes.string,
  }),
};

Spinner.defaultProps = {
  finalProgressValue: 100,
  circleAttributes: {
    radius: 250,
    circleStrokeWidth: 10,
    circleColor: '#e6e6e6',
    circleProgressColor: '#4169E1',
    rotationSpeed: '1s',
  },
};

export default Spinner;
