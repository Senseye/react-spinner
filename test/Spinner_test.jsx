import React from 'react';
import renderer from 'react-test-renderer';

import Spinner from '../src/components/Spinner';

describe('Spinner Component', () => {
  const createSpinnerComponent = props => renderer.create(<Spinner {...props} />);
  const assertComponentSnapshot = component => expect(component.toJSON()).toMatchSnapshot();

  const createComponentsAndAssert = (propsList) => {
    propsList.forEach((props) => {
      const c = createSpinnerComponent(props);
      assertComponentSnapshot(c);
    });
  };

  describe('without optional parameters', () => {
    it('should render without the progress circle and animation tags for no progress', () => {
      const component = createSpinnerComponent({ actualProgress: 0 });
      assertComponentSnapshot(component);
    });

    it('should render with the progress circle and animation tags till completed', () => {
      const progressValues = [
        { actualProgress: 25 },
        { actualProgress: 50 },
        { actualProgress: 75 },
        { actualProgress: 75 },
        { actualProgress: 100 },
        { actualProgress: 105 },
      ];
      createComponentsAndAssert(progressValues);
    });
  });

  describe('with all parameters', () => {
    const props = {
      circleAttributes: {
        radius: 25,
        circleStrokeWidth: 5,
        circleColor: '#ff0000',
        circleProgressColor: '#000',
        rotationSpeed: '0.5s',
      },
      finalProgressValue: 312,
    };

    it('should render without the progress circle and animation tags for no progress', () => {
      const component = createSpinnerComponent({ ...props, actualProgress: 0 });
      assertComponentSnapshot(component);
    });

    it('should render with the progress circle and animation tags till completed', () => {
      const progressProps = [
        { props, actualProgress: 25 },
        { props, actualProgress: 50 },
        { props, actualProgress: 200 },
        { props, actualProgress: 200 },
        { props, actualProgress: 312 },
        { props, actualProgress: 322 },
      ];
      createComponentsAndAssert(progressProps);
    });
  });
});
