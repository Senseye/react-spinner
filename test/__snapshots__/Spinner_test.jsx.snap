// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner Component with all parameters should render with the progress circle and animation tags till completed 1`] = `
<svg
  height={510}
  width={510}
>
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#e6e6e6"
    strokeWidth={10}
  />
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#4169E1"
    strokeDasharray={1570.7963267948965}
    strokeDashoffset={1178.0972450961724}
    strokeWidth={10}
  >
    <animateTransform
      attributeName="transform"
      attributeType="XML"
      dur="1s"
      from="360 255 255"
      repeatCount="indefinite"
      to="0 255 255"
      type="rotate"
    />
  </circle>
  <text
    fontSize={136}
    textAnchor="middle"
    x={255}
    y={300.3333333333333}
  >
    25
  </text>
</svg>
`;

exports[`Spinner Component with all parameters should render with the progress circle and animation tags till completed 2`] = `
<svg
  height={510}
  width={510}
>
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#e6e6e6"
    strokeWidth={10}
  />
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#4169E1"
    strokeDasharray={1570.7963267948965}
    strokeDashoffset={785.3981633974482}
    strokeWidth={10}
  >
    <animateTransform
      attributeName="transform"
      attributeType="XML"
      dur="1s"
      from="360 255 255"
      repeatCount="indefinite"
      to="0 255 255"
      type="rotate"
    />
  </circle>
  <text
    fontSize={136}
    textAnchor="middle"
    x={255}
    y={300.3333333333333}
  >
    50
  </text>
</svg>
`;

exports[`Spinner Component with all parameters should render with the progress circle and animation tags till completed 3`] = `
<svg
  height={510}
  width={510}
>
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#e6e6e6"
    strokeWidth={10}
  />
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#4169E1"
    strokeDasharray={1570.7963267948965}
    strokeDashoffset={-1570.7963267948965}
    strokeWidth={10}
  />
  <text
    fontSize={136}
    textAnchor="middle"
    x={255}
    y={300.3333333333333}
  >
    200
  </text>
</svg>
`;

exports[`Spinner Component with all parameters should render with the progress circle and animation tags till completed 4`] = `
<svg
  height={510}
  width={510}
>
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#e6e6e6"
    strokeWidth={10}
  />
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#4169E1"
    strokeDasharray={1570.7963267948965}
    strokeDashoffset={-1570.7963267948965}
    strokeWidth={10}
  />
  <text
    fontSize={136}
    textAnchor="middle"
    x={255}
    y={300.3333333333333}
  >
    200
  </text>
</svg>
`;

exports[`Spinner Component with all parameters should render with the progress circle and animation tags till completed 5`] = `
<svg
  height={510}
  width={510}
>
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#e6e6e6"
    strokeWidth={10}
  />
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#4169E1"
    strokeDasharray={1570.7963267948965}
    strokeDashoffset={-3330.088212805181}
    strokeWidth={10}
  />
  <text
    fontSize={136}
    textAnchor="middle"
    x={255}
    y={300.3333333333333}
  >
    312
  </text>
</svg>
`;

exports[`Spinner Component with all parameters should render with the progress circle and animation tags till completed 6`] = `
<svg
  height={510}
  width={510}
>
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#e6e6e6"
    strokeWidth={10}
  />
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#4169E1"
    strokeDasharray={1570.7963267948965}
    strokeDashoffset={-3487.1678454846706}
    strokeWidth={10}
  />
  <text
    fontSize={136}
    textAnchor="middle"
    x={255}
    y={300.3333333333333}
  >
    322
  </text>
</svg>
`;

exports[`Spinner Component with all parameters should render without the progress circle and animation tags for no progress 1`] = `
<svg
  height={55}
  width={55}
>
  <circle
    cx={27.5}
    cy={27.5}
    fill="none"
    r={25}
    stroke="#ff0000"
    strokeWidth={5}
  />
  <text
    fontSize={14.666666666666666}
    textAnchor="middle"
    x={27.5}
    y={32.388888888888886}
  >
    0
  </text>
</svg>
`;

exports[`Spinner Component without optional parameters should render with the progress circle and animation tags till completed 1`] = `
<svg
  height={510}
  width={510}
>
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#e6e6e6"
    strokeWidth={10}
  />
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#4169E1"
    strokeDasharray={1570.7963267948965}
    strokeDashoffset={1178.0972450961724}
    strokeWidth={10}
  >
    <animateTransform
      attributeName="transform"
      attributeType="XML"
      dur="1s"
      from="360 255 255"
      repeatCount="indefinite"
      to="0 255 255"
      type="rotate"
    />
  </circle>
  <text
    fontSize={136}
    textAnchor="middle"
    x={255}
    y={300.3333333333333}
  >
    25
  </text>
</svg>
`;

exports[`Spinner Component without optional parameters should render with the progress circle and animation tags till completed 2`] = `
<svg
  height={510}
  width={510}
>
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#e6e6e6"
    strokeWidth={10}
  />
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#4169E1"
    strokeDasharray={1570.7963267948965}
    strokeDashoffset={785.3981633974482}
    strokeWidth={10}
  >
    <animateTransform
      attributeName="transform"
      attributeType="XML"
      dur="1s"
      from="360 255 255"
      repeatCount="indefinite"
      to="0 255 255"
      type="rotate"
    />
  </circle>
  <text
    fontSize={136}
    textAnchor="middle"
    x={255}
    y={300.3333333333333}
  >
    50
  </text>
</svg>
`;

exports[`Spinner Component without optional parameters should render with the progress circle and animation tags till completed 3`] = `
<svg
  height={510}
  width={510}
>
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#e6e6e6"
    strokeWidth={10}
  />
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#4169E1"
    strokeDasharray={1570.7963267948965}
    strokeDashoffset={392.6990816987241}
    strokeWidth={10}
  >
    <animateTransform
      attributeName="transform"
      attributeType="XML"
      dur="1s"
      from="360 255 255"
      repeatCount="indefinite"
      to="0 255 255"
      type="rotate"
    />
  </circle>
  <text
    fontSize={136}
    textAnchor="middle"
    x={255}
    y={300.3333333333333}
  >
    75
  </text>
</svg>
`;

exports[`Spinner Component without optional parameters should render with the progress circle and animation tags till completed 4`] = `
<svg
  height={510}
  width={510}
>
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#e6e6e6"
    strokeWidth={10}
  />
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#4169E1"
    strokeDasharray={1570.7963267948965}
    strokeDashoffset={392.6990816987241}
    strokeWidth={10}
  >
    <animateTransform
      attributeName="transform"
      attributeType="XML"
      dur="1s"
      from="360 255 255"
      repeatCount="indefinite"
      to="0 255 255"
      type="rotate"
    />
  </circle>
  <text
    fontSize={136}
    textAnchor="middle"
    x={255}
    y={300.3333333333333}
  >
    75
  </text>
</svg>
`;

exports[`Spinner Component without optional parameters should render with the progress circle and animation tags till completed 5`] = `
<svg
  height={510}
  width={510}
>
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#e6e6e6"
    strokeWidth={10}
  />
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#4169E1"
    strokeDasharray={1570.7963267948965}
    strokeDashoffset={0}
    strokeWidth={10}
  />
  <text
    fontSize={136}
    textAnchor="middle"
    x={255}
    y={300.3333333333333}
  >
    100
  </text>
</svg>
`;

exports[`Spinner Component without optional parameters should render with the progress circle and animation tags till completed 6`] = `
<svg
  height={510}
  width={510}
>
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#e6e6e6"
    strokeWidth={10}
  />
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#4169E1"
    strokeDasharray={1570.7963267948965}
    strokeDashoffset={-78.53981633974489}
    strokeWidth={10}
  />
  <text
    fontSize={136}
    textAnchor="middle"
    x={255}
    y={300.3333333333333}
  >
    105
  </text>
</svg>
`;

exports[`Spinner Component without optional parameters should render without the progress circle and animation tags for no progress 1`] = `
<svg
  height={510}
  width={510}
>
  <circle
    cx={255}
    cy={255}
    fill="none"
    r={250}
    stroke="#e6e6e6"
    strokeWidth={10}
  />
  <text
    fontSize={136}
    textAnchor="middle"
    x={255}
    y={300.3333333333333}
  >
    0
  </text>
</svg>
`;
